modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java

```diff
@@ -241,6 +250,56 @@ private static List<Jsons.Val> assembleTracks(SearchResultItem event, MediaPacka
         .collect(Collectors.toCollection(ArrayList::new));
   }
 
+  private static List<Jsons.Val> findCaptions(MediaPackage mp) {
+    // We deduplicate the captions by URL and language in case we somehow find
+    // one twice.
+    var captions = new HashMap<String, HashSet<String>>();
+    Arrays.stream(mp.getElements())
+        .filter(element -> {
+          final var isVTT = element.getFlavor().toString().startsWith("captions/vtt")
+                || element.getMimeType().eq("text", "vtt");
+          final var isCorrectType = element.getElementType() == MediaPackageElement.Type.Attachment
+                || element.getElementType() == MediaPackageElement.Type.Track;
+
+          return isVTT && isCorrectType;
+        })
+        .forEach(track -> {
+          final var uri = track.getURI().toString();
+          captions.putIfAbsent(uri, new HashSet<String>());
+
+          // Try to get language from flavor. Otherwise there is an empty list
+          // in the map.
+          final var subflavor = track.getFlavor().getSubtype();
+          if (subflavor.startsWith("vtt+")) {
+            final var suffix = subflavor.substring("vtt+".length());
+            if (suffix.length() > 0) {
+              captions.get(uri).add(suffix);
+            }
+          }
+
+        });
+
+    return captions.entrySet()
+        .stream()
+        .flatMap(e -> {
+          final var languages = e.getValue();
+          final var uri = e.getKey();
+
+          // If this caption URL had no language associated with it, we push a
+          // `null` language to emit this caption at all. Otherwise, all
+          // languages are output as a separate track.
+          if (languages.isEmpty()) {
+            languages.add(null);
+          }
```

Comment:
If `lang` is a set anyway, why add `null` to the set and not just transmit an empty set?

Reply:
I `flatMap` the languages below that and if the set were empty, the subtitle wouldn't be sent at all. I have to admit, I'm not too happy with this whole code, but I didn't find a better way to code this logic.